arangod/Aql/Optimizer/Rule/OptimizerRuleCollectWithIndex.cpp

```diff
@@ -0,0 +1,290 @@
+
+#define LOG_RULE LOG_DEVEL_IF(false)
+
+namespace {
+
+bool isCollectNodeEligible(CollectNode const& cn) {
+  if (cn.hasOutVariable()) {
+    LOG_RULE << "Collect " << cn.id()
+             << " not eligible - it has an out variable";
+    return false;
+  }
+
+  if (not cn.aggregateVariables().empty()) {
+    LOG_RULE << "Collect " << cn.id() << " not eligible - it has aggregations";
+    return false;
+  }
+
+  if (cn.getOptions().disableIndexUsage) {
+    LOG_RULE << "Collect " << cn.id()
+             << " not eligible - index usage explicitly disabled";
+    return false;
+  }
+
+  return true;
+}
+
+bool isIndexNodeEligible(IndexNode const& in) {
+  if (in.hasFilter()) {
+    LOG_RULE << "IndexNode " << in.id()
+             << " not eligible - it has a post filter";
+    return false;
+  }
+  if (in.isLateMaterialized()) {
+    LOG_RULE << "IndexNode " << in.id()
+             << " not eligible - it is in late materialize mode";
+    return false;
+  }
+  if (in.getIndexes().size() != 1) {
+    LOG_RULE << "IndexNode " << in.id()
+             << " not eligible - it uses multiple indexes";
+    return false;
+  }
+  auto index = in.getSingleIndex();
+  if (index->type() != arangodb::Index::TRI_IDX_TYPE_PERSISTENT_INDEX) {
+    LOG_RULE << "IndexNode " << in.id()
+             << " not eligible - only persistent index supported";
+    return false;
+  }
+  if (not in.projections().usesCoveringIndex()) {
+    LOG_RULE << "IndexNode " << in.id()
+             << " not eligible - index does not cover everything";
+    return false;
+  }
+  if (not in.condition()->isEmpty()) {
+    LOG_RULE << "IndexNode " << in.id()
+             << " not eligible - index with condition is not yet supported";
+    return false;
+  }
+  if (index->sparse()) {
+    LOG_RULE << "IndexNode " << in.id()
+             << " not eligible - sparse indexes not yet supported";
+    return false;
+  }
+  if (index->unique()) {
+    LOG_RULE << "IndexNode " << in.id()
+             << " not eligible - optimization not feasible for unique indexes";
+    return false;
+  }
+
+  if (not in.options().ascending) {
+    LOG_RULE << "IndexNode " << in.id()
+             << " not eligible - descending mode not yet supported";
+    return false;
+  }
+
+  // assume there are n documents in the collection and we have
+  // k distinct features. A linear search is in O(n) while a distinct scan
+  // requires O(k log n). So checking for k log n < n, it follows
+  // k / n < 1 / log n, where k / n is precisely the selectivity estimate.
+  double selectivity = index->selectivityEstimate();
+  auto numberOfItems = in.estimateCost().estimatedNrItems;
+
+  double const requiredSelectivity = 1. / log(numberOfItems);
+
+  if (selectivity > requiredSelectivity) {
+    LOG_RULE << "IndexNode " << in.id()
+             << " not eligible - selectivity is too high, actual = "
+             << selectivity << " max allowed = " << requiredSelectivity;
+    return false;
+  }
```

Comment:
I think this might make slightly more sense in a separate function from `isIndexNodeEligible`.